README
NAME
    Sub::Spec::CmdLine - Access Perl subs via command line

VERSION
    version 0.37

SYNOPSIS
    In your module:

     package YourModule;
     our %SPEC;

     $SPEC{foo} = {
         summary => 'Foo!',
         args => {
             arg  => ...,
             arg2 => ...
         },
         ...
     };
     sub foo {
        ...
     }

     ...
     1;

    In your script:

     #!/usr/bin/perl
     use Sub::Spec::CmdLine qw(run);
     run(module=>'YourModule', sub=>'foo');

    In the command-line:

     % script.pl --help
     % script.pl --arg value --arg2 '[an, array, in, yaml, syntax]' ...

    For running multiple subs, in your script:

     use Sub::Spec::CmdLine qw(run);
     run(subcommands => {
         foo => { module=>'YourModule', sub=>'foo'},
         bar => { module=>'YourModule', sub=>'bar'},
         ...
     });

    In the command-line:

     % script.pl --help
     % script.pl --list
     % script.pl foo --help
     % script.pl foo --arg value --arg2 ...
     % script.pl bar --blah ...

DESCRIPTION
    NOTICE: This module and the Sub::Spec standard is deprecated as of Jan
    2012. Rinci is the new specification to replace Sub::Spec, it is about
    95% compatible with Sub::Spec, but corrects a few issues and is more
    generic. "Perinci::*" is the Perl implementation for Rinci and many of
    its modules can handle existing Sub::Spec sub specs. See
    Perinci::CmdLine which supersedes this module.

    This module utilize sub specs (as defined by Sub::Spec) to let your subs
    be accessible from the command-line.

    This can be used to create a command-line application easily. What
    you'll get:

    *   Command-line parsing (currently using Getopt::Long, with some
        tweaks)

    *   Help message (utilizing information from sub specs)

    *   Tab completion for bash

    This module uses Log::Any logging framework. Use something like
    Log::Any::App, etc to see more logging statements for debugging.

    Note: If you use this module, make sure that your sub does not return
    status code above 555, because OS exit code is set to $code-300.

FUNCTIONS
    None of the functions are exported by default, but they are exportable.

  format_result($sub_res[, \%opts]) => TEXT
    Format result from sub into various formats

    Options:

    *   format => FORMAT (optional, default 'text')

        Format can be 'text' (pretty text or nonpretty text), 'pretty'
        (pretty text, generated by Data::Format::Pretty::Console under
        interactive=1), 'nopretty' (also generated by
        Data::Format::Pretty::Console under interactive=0), 'yaml', 'json',
        'php' (generated by PHP::Serialization's serialize()).

    *   default_success_message => STR (optional, default none)

        If output format is text ('text', 'pretty', 'nopretty') and result
        code is 200 and there is no data returned, this
        default_success_message is used. Example: 'Success'.

  run(%args)
    Run subroutine(s) from the command line, which essentially comprises
    these steps:

    *   Parse command-line options in @ARGV (using Sub::Spec::GetArgs::Argv)

        Also, display help using Sub::Spec::To::Text::Usage::spec_to_usage()
        if given '--help' or '-h' or '-?'.

        See Sub::Spec::GetArgs::Argv for details on parsing.

    *   Call sub

    *   Format the return value from sub (using format_result())

    *   Exit with appropriate exit code

        0 if 200, or CODE-300.

    Arguments (* denotes required arguments):

    *   summary => STR

        Used when displaying help message or version.

    *   module => STR

        Currently this must be supplied if you want --version to work, even
        if you use subcommands. --version gets $VERSION from the main
        module. Not required if you specify 'spec'.

    *   sub => STR

        Required if you only want to execute one subroutine. Alternatively
        you can provide multiple subroutines from which the user can choose
        (see 'subcommands').

    *   spec => HASH | CODEREF

        Instead of trying to look for the spec using module and sub, use the
        supplied spec.

    *   help => STRING | CODEREF

        Instead of generating help using spec_to_usage() from the spec, use
        the supplied help message (or help code, which is expected to return
        help text).

    *   subcommands => {NAME => {ARGUMENT=>...}, ...} | CODEREF

        module and sub should be specified if you only have one sub to run.
        If you have several subs to run, assign each of them to a
        subcommand, e.g.:

         summary     => 'Maintain a directory containing git repos',
         module      => 'Git::Bunch',
         subcommands => {
           check   => { },
           backup  => { }, # module defaults to main module argument,
           sync    => { }, # sub defaults to the same name as subcommand name
         },

        Available argument for each subcommand: 'module' (defaults to main
        module argument), 'sub' (defaults to subcommand name), 'summary',
        'help', 'category' (for arrangement when listing commands), 'run',
        'complete_arg', 'complete_args'.

        Subcommand argument can be a code reference, in which case it will
        be called with %args containing: 'name' (subcommand name), 'args'
        (arguments to run()). The code is expected to return structure for
        argument with specified name, or, when name is not specified, a
        hashref containing all subcommand arguments.

    *   run => CODEREF

        Instead of running command by invoking subroutine specified by
        module and sub, run this code instead. Code is expected to return a
        response structure ([CODE, MESSAGE, DATA]).

    *   exit => BOOL (default 1)

        If set to 0, instead of exiting with exit(), return the exit code
        instead.

    *   load => BOOL (default 1)

        If set to 0, do not try to load (require()) the module.

    *   allow_unknown_args => BOOL (default 0)

        If set to 1, unknown command-line argument will not result in fatal
        error.

    *   complete_arg => {ARGNAME => CODEREF, ...}

        Under bash completion, when completing argument value, you can
        supply a code to provide its completion. Code will be called with
        %args containing: word, words, arg, args.

    *   complete_args => CODEREF

        Under bash completion, when completing argument value, you can
        supply a code to provide its completion. Code will be called with
        %args containing: word, words, arg, args.

    *   custom_completer => CODEREF

        To be passed to Sub::Spec::BashComplete's bash_complete_spec_arg().
        This can be used e.g. to change bash completion code (e.g. calling
        bash_complete_spec_arg() recursively) based on context.

    *   dash_to_underscore => BOOL (optional, default 0)

        If set to 1, subcommand like a-b-c will be converted to a_b_c. This
        is for convenience when typing in command line.

    *   undo => BOOL (optional, default 0)

        If set to 1, --undo and --undo-dir will be added to command-line
        options. --undo is used to perform undo: -undo and -undo_data will
        be passed to subroutine, an error will be thrown if subroutine does
        not have 'undo' features. --undo-dir is used to set location of undo
        data (default ~/.undo; undo directory will be created if not exists;
        each subroutine will have its own subdir here).

    run() can also perform completion for bash (if Sub::Spec::BashComplete
    is available). To get bash completion for your perlprog, just type this
    in bash:

     % complete -C /path/to/perlprog perlprog

    You can add that line in bash startup file (~/.bashrc, /etc/bash.bashrc,
    etc).

FAQ
  How does Sub::Spec::CmdLine compare with other CLI-app frameworks?
    Differences: Sub::Spec::CmdLine is part of a more general subroutine
    metadata framework. Aside from a command-line app, your sub spec is also
    usable for other stuffs, like creating REST API's, remote subroutines,
    or documentation. Sub::Spec::CmdLine is not OO and does not offer
    plugins (as of now).

    Pros: App::Cmd and App::Rad currently does not offer bash completion
    feature. Sub::Spec::CmdLine offers passing arguments as YAML.

    Cons: inadequate documentation/tutorial, no configuration file support
    yet (coming soon).

  Why is nonscalar arguments parsed as YAML instead of JSON/etc?
    I think YAML is nicer in command-line because quotes are optional in a
    few places:

     $ cmd --array '[a, b, c]' --hash '{foo: bar}'

    versus:

     $ cmd --array '["a", "b", "c"]' --hash '{"foo": "bar"}'

    Though YAML requires spaces in some places where JSON does not. A flag
    to parse as JSON can be added upon request.

SEE ALSO
    App::Cmd, App::Rad

    Sub::Spec

    MooseX::Getopt

AUTHOR
    Steven Haryanto <stevenharyanto@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by Steven Haryanto.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Sub::Spec::CmdLine

0.37     2012-01-20

         No functional changes. Mark deprecation.


0.36     2011-08-25

         No functional changes. gen_usage() split to its own dist:
         Sub-Spec-To-Text. Update to Sub::Spec 0.15.


0.35     2011-08-11

         Uses Data::Format::Pretty for result formatting. JSON output is now
         pretty-printed by default.


0.34     2011-08-03

         No functional changes. Fixed test script to adjust to new
         Object::BlankStr [CT]. Updated dependencies.


0.33     2011-07-26

         [INCOMPATIBLE CHANGES]

         Refactoring: move parse_argv() to
         Sub::Spec::GetArgs::Argv::get_args_from_argv().

         [ENHANCEMENTS]

         Localize Getopt::Long::Configure() effect.


0.32     2011-07-10

         [BUG FIXES]

         * runpm: fix -I.


0.31     2011-07-10

         [ENHANCEMENTS]

         * bin/runpm: add --help (-h) and -I (--library) options.

         * run(): add dash_to_underscore option to allow subcommands using
           dashes instead of underscores.


0.30     2011-06-05

         [INCOMPATIBLE CHANGES]

         * To reduce the trap of implicit YAML parsing, parsing is only done for
           nonscalar arguments.


0.29     2011-06-01

         No functional changes.

         [ETC]

         - Update to latest version of Sub::Spec::Runner.


0.28     2011-04-14

         [BUG FIXES]

         - Remove extra name line in usage message.


0.27     2011-04-14

         [ENHANCEMENTS]

         - run(): For convenience, allow general options (e.g. --version,
           --help, --json, etc) to be specified after subcommand name, if it
           doesn't collide with spec arg

         [ETC]

         - Replace Clone::Any with Data::Clone (Clone::Any fallbacks to Storable
           which doesn't support storing Regexp, so tests will fail if Clone is
           not available).


0.26     2011-03-04

         Require Sub::Spec::Runner 0.09.


0.25     2011-02-28

         On non-success response from sub, print output (error message) even if
         spec enables cmdline_suppress_output.


0.24     2011-02-28

         Now based on Sub::Spec::Runner for dependency checking before run.


0.23     2011-02-23

         Fix dependency (Sub::Spec version).

         Don't add error message when Getopt::Long already prints to stderr.


0.22     2011-02-22

         Use Sub::Spec::Utils.


0.21     2011-02-21

         Add some more documentation.

         Requires recent versions of Test::More & Sub::Spec::BashComplete.


0.20     2011-02-18

         Escape completion output since BashComplete now doesn't.

         Added some doc.


0.19     2011-02-18

         Provide runpm script.


0.18     2011-02-18

         run(): More fixes for completion.

         spec: cmdline_aliases is renamed to arg_aliases.

         Fix test [CT].


0.17     2011-02-17

         run(): Fixed stupid parenthesis bug.


0.16     2011-02-17

         run(): A couple more fixes for bash completion.

         parse_argv(): Also add --nofoo for bool cmdline_aliases.


0.15     2011-02-17

         Yet more fixes for bash completion.

         Fixed: parse_argv(): do not convert all remaining $argv to YAML so they
         can be reused (e.g. parse_argv'd) again

         run(): add option 'custom_completer' to pass to
         Sub::Spec::BashComplete.


0.14     2011-02-17

         Fixes for bash completion.


0.13     2011-02-17

       [INCOMPATIBLE CHANGES]

         run(): rename 'require' argument to 'load'.

       [IMPROVEMENTS]

         parse_argv(): Respect arg spec: 'cmdline_aliases' for specifying
         option aliases.

         run(): Allow 'subcommands' argument to be coderef.

         run(): Pass more info (arguments) to coderef 'help' argument.

         run(): Add option 'allow_unknown_args' for non-strict argv parsing.

         gen_usage(): Add option: 'options_name'.


0.12     2011-02-16

         Fixes for bash completion.

         Add arguments to run(): help, spec, run, per-subcommand category.

         Other bug fixes.


0.11     2011-02-16

         Add support for bash completion (requires Sub::Spec::BashComplete).


0.10     2011-02-16

         Add strict=>0 option in parse_argv(), for use by
         Sub::Spec::BashComplete.

         Some small refactoring.


0.09     2011-02-15

         Add --version.


0.08     2011-02-15

         Fixes for subcommands support.

         Respect spec clause: cmdline_suppress_output.


0.07     2011-02-11

         Formatting fix: missing newline in error message.


0.06     2011-02-11

         Add 'cmd' option to gen_usage().

         Formatting tweaks.


0.05     2011-02-11

         Remove 'required_args'.


0.04     2011-02-10

         Formatting fixes.


0.03     2011-02-10

         Formatting tweaks.

         Fix schema clause ('in', not 'choices').


0.02     2011-02-09

         Incompatible change: rename arg_order to arg_pos.

         Documentation.


0.01     2011-02-09

         First release.


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Access Perl subs via command line",
  "dist_author" => [
    "Steven Haryanto <stevenharyanto\@gmail.com>"
  ],
  "dist_name" => "Sub-Spec-CmdLine",
  "dist_version" => "0.37",
  "license" => "perl",
  "module_name" => "Sub::Spec::CmdLine",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Capture::Tiny" => 0,
    "Data::Clone" => 0,
    "Data::Format::Pretty" => 0,
    "Data::Format::Pretty::Console" => 0,
    "Data::Format::Pretty::JSON" => 0,
    "Data::Format::Pretty::PHPSerialization" => 0,
    "Data::Format::Pretty::YAML" => 0,
    "Data::Sah" => 0,
    "JSON" => 0,
    "List::MoreUtils" => 0,
    "Log::Any" => 0,
    "Sub::Spec" => "0.15",
    "Sub::Spec::BashComplete" => "0.16",
    "Sub::Spec::GetArgs::Argv" => "0.05",
    "Sub::Spec::Object" => "0.02",
    "Sub::Spec::Runner" => "0.21",
    "Sub::Spec::To::Text::Usage" => 0,
    "Test::More" => "0.96",
    "perl" => "5.010000"
  },
  "script_files" => [
    "bin/runpm"
  ]
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
